### PR TITLE
Initial JupyterBook ingest feature

### PR DESCRIPTION
**Pull request by jonathansick**
This is a _big_ PR that essentially gets Astropy Librarian to a state where we can use it to populate guide and tutorial content from the Learn Astropy front end.

The main feature is the new capability to ingest JupyterBooks, which are our primary format for writing *guides*.

This PR also improves the existing ingest workflow for tutorial notebooks by leveraging [Pydantic models](https://pydantic-docs.helpmanual.io) — Pydantic is like dataclasses but with enhanced validation of datasets and extra control over how data is exported and serialized.

The Algolia record model for both tutorials and guides is based on the same base `AlgoliaRecord` type, though each content type has slight differences in what "optional" attributes are available.

This work implements a [typer](https://typer.tiangolo.com) based CLI to run workflows through subcommands, such as `astropylibrarian index tutorial` or `astropylibrarian index guide` or `astropylibrarian delete`. The README includes the help output from the CLI.

This PR updates the tutorial index workflow to work a "new" HTML structure for pages where instead of `<div class="section">` elements containing each section, the semantic `<section>` element is used instead. This new structure is appearing in our current learn.astropy.org tutorials. The old structure is still seen in the CCD processing guide. This PR enables the section iterator to handle both.

Finally, this PR introduces a feature for expiring old records (which can happen if we re-index a guide/tutorial, and that content no longer has a section, or section changes title, etc.). This is done by creating a unique `index_epoch` key for each indexing even. After indexing, we look for records corresponding to that `root_url` that have a _different_ `index_epoch` value than the records we just saved.

### CODE REVIEW

File: astropylibrarian/algolia/client.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+
+class MockAlgoliaIndex(BaseAlgoliaIndex):
+    """A mock Algolia index client.
+
+    Use this client as a drop-in replaceemnt to `AlgoliaIndex` in situations
+    where you do not want to make real network requests to Algolia, such as in
+    testing or in dry-run applications. This client keeps tracks of what calls
+    it would have made so that you can understand how your application uses
+    Algolia.
```
Comment (adrn) [inline astropylibrarian/algolia/client.py:108]: Phrasing here is a little confusing

File: astropylibrarian/algolia/records.py [hunk 2]
```diff
@@ -1,110 +1,314 @@
...
+
+    guide = "guide"
+
+    example = "example"
+
+    documentation = "documentation"
+
+
+class AlgoliaRecord(BaseModel):
+    """A Pydantic model for an Learn Astropy record in Algolia."""
```
Comment (adrn) [inline astropylibrarian/algolia/records.py:40]:
```suggestion
    """A Pydantic model for a Learn Astropy record in Algolia."""
```
Fitting for this line that this is a *very* pedantic suggestion 😅

File: astropylibrarian/algolia/records.py [hunk 3]
```diff
@@ -1,110 +1,314 @@
...
-    @property
-    def object_id(self) -> str:
-        """The objectID of the record.
+    importance: int = Field(
+        description="The importance of the record, corresponding to the "
+        "hierarchical section level"
+    )
 
-        This is computed based on the URL and section heading hierarchy.
+    content: str = Field(description="The plain text content of the record.")
```
Comment (adrn) [inline astropylibrarian/algolia/records.py:108]: Just to be clear: here "plain text" is actually HTML?

File: astropylibrarian/algolia/records.py [hunk 4]
```diff
@@ -1,110 +1,314 @@
...
+                tutorial.keywords, "task"
+            ),
+            "science_keywords": keyworddb.filter_keywords(
+                tutorial.keywords, "science"
+            ),
+        }
+        for i, heading in enumerate(section.headings):
+            kwargs[f"h{i+1}"] = heading
+        if tutorial.images:
+            kwargs["thumbnail_url"] = tutorial.images[0]
```
Comment (adrn) [inline astropylibrarian/algolia/records.py:249]: Add a TODO comment here to note that we might want to make this customizable via notebook metadata? Also will this fail if there are no images?

File: astropylibrarian/algolia/records.py [hunk 5]
```diff
@@ -1,110 +1,314 @@
...
+    def from_section(
+        cls,
+        *,
+        site_metadata: JupyterBookMetadata,
+        page: JupyterBookPage,
+        section: Section,
+        index_epoch: str,
+    ) -> GuideRecord:
+        if page.image_urls:
+            thumbnail_url = page.image_urls[0]
```
Comment (adrn) [inline astropylibrarian/algolia/records.py:291]: Similarly here (to comment above)?

File: astropylibrarian/reducers/tutorial.py [hunk 6]
```diff
@@ -157,6 +165,32 @@ def _is_ignored_section(self, section: Section) -> bool:
...
+                tutorial=self,
+                section=section,
+                keyworddb=keyworddb,
+                index_epoch=index_epoch,
+            )
+
+    def iter_algolia_objects(
+        self, *, index_epoch: str
+    ) -> Iterator[Dict[str, Any]]:
+        """Iterate over all objects that are extratable from the tutorial in
```
Comment (adrn) [inline astropylibrarian/reducers/tutorial.py:182]:
```suggestion
        """Iterate over all objects that are extractable from the tutorial in
```

**Review by adrn — APPROVED** [review]
Thanks for this @jonathansick!

I admit I didn't do an incredibly careful review of the code, but the overall structure looks good to me, and the code is clean, readable, tested, and documented -- so I vote (after fixing the very few minor suggestions) that we merge this 🎉

File: astropylibrarian/algolia/records.py (hunk 3 above)
Comment (jonathansick) [reply astropylibrarian/algolia/records.py:108]: It really is the unstyled text; effectively all the content that goes inside HTML tags. How we get this for web pages is with [lxml.html's text_content() method](https://lxml.de/lxmlhtml.html). The reason we only want to track plain text content is that that's what Algolia is searching against, so we don't want all that markup in this instance.

**Comment by jonathansick** [comment]
Thanks @adrn !

**Pull request merged**